File: 3D Objects/snowFace/Graph/ScriptTemplate/Anim3DControl/CGAnimatorController.js
/**
 * @file CGAnimatorController.js
 * @author xuyuan
 * @date 2021/10/14
 * @brief CGAnimatorController.js
 * @copyright Copyright (c) 2021, ByteDance Inc, All Rights Reserved
 */

const Amaz = effect.Amaz;
const {BaseNode} = require("./BaseNode");

class CGAnimatorController extends BaseNode {
    constructor() {
        super();
        this.component = null;
        this.animationList = new Array();
        this.animationSize = 0;
        this.currentClip = null;
        this.loops = 0;
        this.currentLoop = 0;
        this.infinity = false;
        this.errorConfig = false;
        this.stayLastFrame = false;
        this.finish = false;
        this.sys = null;
        this.state = '';
    }

    beforeStart(sys) {
        this.sys = sys;
        this.component = this.inputs[4]();
        if (!this.component || false === this.component.isInstanceOf("Animator")) {
            this.errorConfig = true;
            return;
        }
        this.loops = this.inputs[6]();
        if (this.loops === 0) {
            console.error("animation loops 0 times !!");
            this.errorConfig = true;
            return;
        }
        if (this.loops === -1) {
            this.infinity = true;
        }
        this.stayLastFrame = this.inputs[7]();
        this.animationList = this.component.animations;
        this.animationSize = this.animationList.size();
        if (this.animationSize < 1) {
            console.error("animations size is 0 !!");
            this.errorConfig = true;
            return;
        }
        let chooseIndex = this.inputs[5]();
        if (chooseIndex >= this.animationSize || chooseIndex < 0 ) {
            console.error("animation chooseIndex error !!");
            this.errorConfig = true;
            return;
        }
        let chooseAnim = this.animationList.get(chooseIndex);
        if (!chooseAnim ) {
            console.error("get chosen animation error !!");
            this.errorConfig = true;
            return;
        }
        this.currentClip = chooseAnim.getClip("", this.component);
        if (!this.currentClip) {
            console.error("get clip from animation error !!");
            this.errorConfig = true;
            return;
        }
        sys.script.removeScriptListener(this.currentClip, Amaz.AnimazEventType.ANIM_END, "onCallBack", sys.script);
        sys.script.addScriptListener(this.currentClip, Amaz.AnimazEventType.ANIM_END, "onCallBack", sys.script);
    }

    execute(index) {
        if (this.nexts[0]) {
            this.nexts[0]();
        }
        if (this.errorConfig) {
            return;
        }
        if (this.finish === true && index !== 0) {
            return;
        }
        if (this.component.entity.visible === false) {
            this.component.entity.visible = true
        }
        if (index === 0) {
            this.finish = false;
            this.currentLoop = 0;
            if (this.sys) {
                this.beforeStart(this.sys);
            }
            this.state = 'play';
            this.component.resumeAnimator();
            this.component.stopAllAnimations();
            let clipPlaySpeed = this.currentClip.getSpeed();
            this.component.schedule(this.currentClip, 1, clipPlaySpeed);
            if (this.nexts[1]) {
                this.nexts[1]();
            }
        } else if (index === 1) {
            this.component.stopAllAnimations();
            this.state = 'stop';
            if (this.nexts[2]) {
                this.nexts[2]();
            }
        } else if (index === 2) {
            this.state = 'pause';
            this.component.pauseAnimator();
        } else if (index === 3) {
            this.state = 'resume';
            this.component.resumeAnimator();
        }
    }

    onCallBack(sys, clip, eventType) {
        if (eventType === Amaz.AnimazEventType.ANIM_END) {
            if (clip.equals(this.currentClip) && this.state !== 'stop') {
                this.currentLoop = this.currentLoop + 1;
                if (this.currentLoop >= this.loops && false === this.infinity) {
                    if (this.stayLastFrame) {
                        this.component.pauseAnimator();
                    } else {
                        this.component.stopAllAnimations();
                        this.component.entity.visible = false;
                    }
                    if (this.nexts[3]) {
                        this.nexts[3]();
                    }
                    this.finish = true;
                } else {
                    let clipPlaySpeed = this.currentClip.getSpeed();
                    this.component.schedule(this.currentClip, 1, clipPlaySpeed);
                }
            }
        }
    }

    onDestroy(sys) {
        sys.script.removeScriptListener(this.currentClip, Amaz.AnimazEventType.ANIM_END, "onCallBack", sys.script);
    }
};

exports.CGAnimatorController = CGAnimatorController;

File: 3D Objects/snowFace/Graph/ScriptTemplate/Anim3DControl/CGAnimatorController.lua
--
-- Created by IntelliJ IDEA.
-- User: xuyuan
-- Date: 2021/5/31
-- Time: 6:42
-- To change this template use File | Settings | File Templates.
--

local CGAnimatorController = CGAnimatorController or {}
CGAnimatorController.__index = CGAnimatorController

function CGAnimatorController.new()
    local self = setmetatable({}, CGAnimatorController)
    self.inputs = {}
    self.outputs = {}
    self.nexts = {}
    self.component = nil
    self.animation = {}
    self.clip = {}
    self.loops = 0
    self.curLoop = {}
    self.stayLastFrame = false
    self.duration = {}
    self.curTime = 0
    self.lastLoopTime = {}
    self.playState = nil
    self.size = 0
    self.enable = false
    return self
end

function CGAnimatorController:setNext(index, func)
    self.nexts[index] = func
end

function CGAnimatorController:setInput(index, func)
    self.inputs[index] = func
end

function CGAnimatorController:getOutput(index)
    return self.outputs[index]
end

function CGAnimatorController:update(sys, dt)
    if self.component == nil or self.enable == false then
        return
    end
    if self.playState == "play" then
        self.curTime = self.curTime + dt
        for i = 1, self.size do
            if self.curTime - self.lastLoopTime[i] >= self.duration[i] then
                self.component:schedule(self.clip[i], 1, 1)
                self.curLoop[i] = self.curLoop[i] + 1
                self.lastLoopTime[i] = self.curTime
            end
        end
    elseif self.playState == "stop" then
        self.curTime = 0
        for i = 1, self.size do
            self.lastLoopTime[i] = 0
            self.curLoop[i] = 0
        end
    end
    local cnt = 0
    for i = 1, self.size do
        if self.curLoop[i] >= self.loops then
            cnt = cnt + 1
        end
    end
    if cnt == self.size then
        self.component:stopAllAnimations()
        if self.stayLastFrame == false then
            self.component.entity.visible = false
        end
    end
end

function CGAnimatorController:execute(index)
    if self.inputs[4] == nil then
        return
    end
    self.component = self.inputs[4]()
    if self.inputs[5] == nil then
        return
    end
    local animations = self.component.animations
    self.size = animations:size()
    if self.size < 1 then
        Amaz.LOGE("ERROR: ", "animations size is 0 !!")
        return
    end
    for i = 1, self.size do
        self.animation[i] = animations:get(i-1)
        self.duration[i] = self.animation[i].duration
        self.clip[i] = self.animation[i]:getClip("", self.component)
        if self.clip[i] == nil then
            Amaz.LOGE("ERROR: ", "got a nil clip !!")
            return
        end
    end
    self.loops = self.inputs[5]()
    self.stayLastFrame = self.inputs[6]()
    if self.component.entity.visible == false then
        self.component.entity.visible = true
    end
    if index == 0 then
        self.curTime = 0
        self.lastTime = 0
        self.component:stopAllAnimations()
        for i = 1, self.size do
            self.curLoop[i] = 0
            self.lastLoopTime[i] = 0
            self.component:schedule(self.clip[i], 1, 1)
        end
        self.playState = "play"
        self.enable = true
        if self.nexts[1] then
            self.nexts[1]();
        end
    elseif index == 1 then
        self.component:stopAllAnimations()
        self.playState = "stop"
        self.enable = false
        if self.nexts[2] then
            self.nexts[2]()
        end
    elseif index == 2 then
        self.component:pauseAnimator()
        self.playState = "pause"
        self.enable = false
    elseif index == 3 then
        self.component:resumeAnimator()
        self.playState = "play"
        self.enable = true
    end
    if self.nexts[0] then
        self.nexts[0]()
    end
end

return CGAnimatorController
